packages/dds/tree/src/core/rebase/rebaser.ts

```diff
@@ -3,223 +3,139 @@
- * @alpha
+ * Handles the rebasing of commits in the commit graph and their changes
  */
-export type RevisionTag = Brand<number, "rebaser.RevisionTag">;
-
-/**
- * A collection of branches which can rebase changes between them.
- *
- * @sealed
- */
-export class Rebaser<TChangeRebaser extends ChangeRebaser<any>> {
-	private lastRevision = 0;
-
-	private makeRevision(): RevisionTag {
-		this.lastRevision++;
-		return brand(this.lastRevision);
-	}
-
-	public readonly empty: RevisionTag = brand(0);
-
+export class Rebaser<TChange> {
 	/**
-	 * All the actual state needed to do the rebases.
-	 *
-	 * Source and destination can both walk this to find common ancestor,
-	 * then rebase across using changes found on walk.
+	 * @param changeRebaser - the change rebaser responsible for rebasing the changes in the commits given to this rebaser
 	 */
-	private readonly revisionTree: Map<
-		RevisionTag,
-		{ before: RevisionTag; change: ChangesetFromChangeRebaser<TChangeRebaser> }
-	> = new Map();
-
-	public constructor(public readonly rebaser: TChangeRebaser) {
-		// TODO
-	}
-
-	/**
-	 * Rebase `changes` from being applied to the `from` state to able to be applied to the `to` state.
-	 * @returns a RevisionTag for the state after applying changes to `to`, and the rebased changes themselves.
-	 */
-	public rebase(
-		changes: ChangesetFromChangeRebaser<TChangeRebaser>,
-		from: RevisionTag,
-		to: RevisionTag,
-	): [RevisionTag, ChangesetFromChangeRebaser<TChangeRebaser>] {
-		const over = this.getResolutionPath(from, to);
-		const finalChangeset: ChangesetFromChangeRebaser<TChangeRebaser> = this.rebaser.rebase(
-			changes,
-			over,
-		);
-		const newRevision = this.makeRevision();
-		this.revisionTree.set(newRevision, {
-			before: to,
-			change: finalChangeset,
-		});
-		return [newRevision, finalChangeset];
-	}
+	public constructor(public readonly changeRebaser: ChangeRebaser<TChange>) {}
 
 	/**
-	 * Modifies `anchors` to be valid at the destination.
+	 * Rebases a branch (the source) onto another branch (the target).
+	 *
+	 * Given a commit as the "head", a branch is defined as the ancestry path of that head commit.
+	 * The source and target branch must share an ancestor. Respectively, they must not contain any commits
+	 * in their path with duplicate tags, or this function will have undefined behavior. If there are any commits
+	 * in `source` that have the same tag as any commits in `target`, then this function will treat each pair as
+	 * semantically equivalent (i.e. one is a pre-existing rebase of the other) and will attempt to minimize the amount
+	 * of rebasing performed for those changes across the branches.
+	 * @param source - the head of a branch
+	 * @param target - the head of branch to rebase `source` onto
+	 * @param base - the commit in the target branch up to which to perform the rebase. Defaults to `target`. Source will be
+	 * rebased up to at least this commit, but it may rebase farther if it is trivial to do so.
```

Comment:
Why do we need this parameter?

Reply:
So that we can do the "catchup optimization" (where the source branch gets rebased even farther forward than requested because its tags align with the next _n_ tags on the target branch). We need to know both the base, which is the actual commit we want to rebase up to, and the target branch, which might contain more commits past the base. There's no way to walk forwards from base since commits don't have pointers to their children; we have to get a point farther ahead and then walk back and see what lines up.